tests/unit/capi/GEOSClipByRectTest.cpp

```diff
@@ -136,7 +136,7 @@ template<> template<> void object::test<10>
 {
     geom1_ = GEOSGeomFromWKT("POLYGON((0 0, 0 30, 30 30, 30 0, 0 0),(10 10, 20 10, 20 20, 10 20, 10 10))");
     geom2_ = GEOSClipByRect(geom1_, 10, 10, 20, 20);
-    isEqual(geom2_, "POLYGON EMPTY");
+    isEqual(geom2_, "POLYGON((10 20, 20 20, 20 10, 10 10, 10 20), (10 10, 20 10, 20 20, 10 20, 10 10))");
```

Comment:
This is way worse than before... the result is a hole exactly on the shell. Which is, yes, empty, but far less legibly than what came before. Is the juice worth the squeeze here?

Reply:
@pramsey thank you for review.

> This is way worse than before... the result is a hole exactly on the shell. Which is, yes, empty, but far less legibly than what came before. Is the juice worth the squeeze here?

right, with these changes there are plenty of cases when geometry becomes invalid after clip whereas with current implementation it's fine but all these geometries are possible to fix with ST_MakeValid while the example I provided in description is unrepairable because polygon is constructed from wrong part of rectangle.

That is what happens when subdivision is done using current ST_ClipByBox2D: 
![Screenshot_20220525_002510](https://user-images.githubusercontent.com/45686940/170135711-77a27046-5802-4e7b-b87d-e2f47cd7ecd0.png)

So I would prefer to have more cases where invalid geometries are produces but these geometries are repairable over function that almost always ok but sometimes outputs completely broken geometry.

For me it looks like making ST_ClipByBox2D use OverlayNG like @dr-jts suggested is the best option here.

Reply:
Yes, it seems validity is the highest order concern. Although aren't we already using OverlayNG in subdivide?

Reply:
Yes, OverlayNG is already used in subdivide. I made it to use ST_ClipByBox2D for my testing purposes, it does plenty of clip operations which makes it easier to find these kind of issues.